x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/action/realm/TransportClearRealmCacheActionTests.java

```diff
@@ -0,0 +1,179 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.action.realm;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.security.action.realm.ClearRealmCacheRequest;
+import org.elasticsearch.xpack.core.security.action.realm.ClearRealmCacheResponse;
+import org.elasticsearch.xpack.core.security.authc.Realm;
+import org.elasticsearch.xpack.core.security.authc.RealmConfig;
+import org.elasticsearch.xpack.core.security.authc.support.CachingRealm;
+import org.elasticsearch.xpack.security.authc.AuthenticationService;
+import org.elasticsearch.xpack.security.authc.Realms;
+import org.junit.Before;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static org.hamcrest.Matchers.notNullValue;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class TransportClearRealmCacheActionTests extends ESTestCase {
+
+    private AuthenticationService authenticationService;
+    private TransportClearRealmCacheAction action;
+    private TestCachingRealm nativeRealm;
+    private TestCachingRealm fileRealm;
+
+    @Before
+    public void setup() {
+        nativeRealm = mockRealm("native");
+        fileRealm = mockRealm("file");
+        final Realms realms = mockRealms(List.of(nativeRealm, fileRealm));
+
+        authenticationService = mock(AuthenticationService.class);
+        action = mockClearRealmCacheAction(realms, authenticationService);
+    }
+
+    public void testSingleUserCacheCleanupForAllRealms() {
+
+        final String user = "test";
+
+        // When no realm is specified we should clear all realms.
+        // This is equivalent to using a wildcard (*) instead of specifying realm name in query.
+        final String[] realmsToClear = new String[] {};
+        final String[] usersToClear = new String[] { user };
+        ClearRealmCacheRequest.Node clearCacheRequest = mockClearCacheRequest(realmsToClear, usersToClear);
+
+        ClearRealmCacheResponse.Node response = action.nodeOperation(clearCacheRequest, mock(Task.class));
+        assertThat(response.getNode(), notNullValue());
+
+        // We expect that caches of all realms are cleared for the given user,
+        // including last successful cache in the authentication service.
+        verify(fileRealm).expire(user);
+        verify(nativeRealm).expire(user);
+        verify(authenticationService).expire(user);
```

Comment:
We could also assert `expireAll` is not called.

Reply:
Makes total sense. Will do.